### PR TITLE
Add support for parser value_set to eBPF backend

### PR DESCRIPTION
**Pull request by tatry**
This PR adds the support for parser value_set to eBPF backend.

Co-authored-by: Tomasz Osiński <tomasz@opennetworking.org>
Co-authored-by: Mateusz Kossakowski <mateusz.kossakowski@orange.com>

### CODE REVIEW

File: backends/ebpf/ebpfParser.h [hunk 1]
```diff
@@ -86,15 +87,21 @@ class EBPFParser : public EBPFObject {
...
+
     explicit EBPFParser(const EBPFProgram* program, const IR::ParserBlock* block,
                         const P4::TypeMap* typeMap);
     virtual void emitDeclaration(CodeBuilder* builder, const IR::Declaration* decl);
     void emit(CodeBuilder* builder);
     virtual bool build();
 
-    virtual void emitTypes(CodeBuilder* builder) { (void) builder; }
-    virtual void emitValueSetInstances(CodeBuilder* builder) { (void) builder; }
+    virtual void emitTypes(CodeBuilder* builder);
```
Comment (mbudiu-vmw) [inline backends/ebpf/ebpfParser.h:98]: So this works with the standard ebpf backend too. Thank you.

File: backends/ebpf/ebpfTable.cpp [hunk 2]
```diff
@@ -872,4 +872,169 @@ void EBPFCounterTable::emitTypes(CodeBuilder* builder) {
...
+                    "Size must be a positive value less than 2^32, got %1% entries", pvs->size);
+    } else {
+        ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                "Size of value_set must be know at compilation time: %1%", pvs->size);
+    }
+
+    // validate type
+    if (pvs->elementType->is<IR::Type_Bits>() || pvs->elementType->is<IR::Type_Tuple>()) {
+        // no restrictions
+    } else if (pvs->elementType->is<IR::Type_Name>()) {
```
Comment (mbudiu-vmw) [inline backends/ebpf/ebpfTable.cpp:899]: In general you can do `typeMap->getType(pvs->elementType, true)`.
This should cover all cases, including nested typedefs, etc.

File: backends/ebpf/ebpfTable.cpp [hunk 3]
```diff
@@ -872,4 +872,169 @@ void EBPFCounterTable::emitTypes(CodeBuilder* builder) {
...
+    } else {
+        ::error(ErrorType::ERR_UNSUPPORTED,
+                "Unsupported type with value_set: %1%", pvs->elementType);
+    }
+
+    keyTypeName = "struct " + keyTypeName;
+}
+
+void EBPFValueSet::emitTypes(CodeBuilder* builder) {
+    if (pvs->elementType->is<IR::Type_Name>()) {
```
Comment (mbudiu-vmw) [inline backends/ebpf/ebpfTable.cpp:922]: Same here, use the typeMap.

File: backends/ebpf/ebpfTable.cpp [hunk 4]
```diff
@@ -872,4 +872,169 @@ void EBPFCounterTable::emitTypes(CodeBuilder* builder) {
...
+    builder->blockStart();
+
+    auto fieldEmitter = [builder](const IR::Type* type, cstring name){
+        auto etype = EBPFTypeFactory::instance->create(type);
+        builder->emitIndent();
+        etype->declare(builder, name, false);
+        builder->endOfStatement(true);
+    };
+
+    if (pvs->elementType->is<IR::Type_Bits>()) {
```
Comment (mbudiu-vmw) [inline backends/ebpf/ebpfTable.cpp:945]: Here as well it is safest to use the type from the typeMap.

File: backends/ebpf/ebpfTable.cpp [hunk 5]
```diff
@@ -872,4 +872,169 @@ void EBPFCounterTable::emitTypes(CodeBuilder* builder) {
...
+            builder->appendLine(",");
+        }
+    }
+
+    builder->blockEnd(false);
+    builder->endOfStatement(true);
+
+    // init other bigger fields
+    for (unsigned int i = 0; i < fieldNames.size(); i++) {
+        if (fieldNames.at(i).second->is<IR::Type_Bits>()) {
```
Comment (mbudiu-vmw) [inline backends/ebpf/ebpfTable.cpp:1012]: why do a separate loop? Can't this case be handled in the above loop?

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: backends/ebpf/ebpfTable.cpp (hunk 5 above)
Comment (tatry) [reply backends/ebpf/ebpfTable.cpp]: I've changed the way that key is initialized, so now it is one loop.

File: backends/ebpf/ebpfTable.cpp (hunk 2 above)
Comment (tatry) [reply backends/ebpf/ebpfTable.cpp]: `getType` returns instance of `IR::Type_Type`. This is intended behavior?

**Comment by tatry** [comment]
@mbudiu-vmw I think I addressed all the comments, thank you for your review.

File: backends/ebpf/ebpfTable.cpp (hunk 2 above)
Comment (mbudiu-vmw) [reply backends/ebpf/ebpfTable.cpp]: Use getTypeType

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**